### ISSUE TITLE
URL in Metadata will encodeURLComponent unexpectedly. Url query `&` will encode into `&amp;` automatically

### ISSUE THREAD

**Issue by Innei** [vercel/next.js#52598]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.5.0: Thu Jun  8 22:21:34 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T8112
    Binaries:
      Node: 18.16.1
      npm: 8.19.4
      Yarn: 1.22.19
      pnpm: 8.6.2
    Relevant Packages:
      next: 13.4.9
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.6
    Next.js Config:
      output: standalone
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App Router, Metadata (metadata, generateMetadata, next/head)

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/Innei/nextjs-metadata-url-encode-reproduction

### To Reproduce

1. `next dev`
2. Open the home page's source view.

### Describe the Bug

Next.js metadata url will encodeURLComponent unexpectedly.

```ts

export const generateMetadata = async (): Promise<Metadata> => {
  const ogUrl = new URL('http://localhost:3000/og.png')

  ogUrl.searchParams.set('title', 'Nextjs')
  ogUrl.searchParams.set('description', 'description - Nextjs')

  return {
    title: 'Nextjs',
    openGraph: {
      images: ogUrl,
    },
  }
}
```

and generate the code of `<meta>` tag.

```
<meta property="og:image" content="http://localhost:3000/og.png?title=Nextjs&amp;description=description+-+Nextjs"/>
```

The `&` will encode to `&amp;` unexpected.

### Expected Behavior

No need to automatically encode HTML characters.

<sub>[NEXT-2729](https://linear.app/vercel/issue/NEXT-2729/url-in-metadata-will-encodeurlcomponent-unexpectedly-url-query-and)</sub>

**Comment by joaolandino** [comment]
Same here 🚨

**Comment by patricktansg** [comment]
any updates for this issues?
Meanwhile, when will we able to inject inline script into head when using App Directory?

**Comment by punarinta** [comment]
Does anyone know if there's a workaround for this problem?

**Comment by woodbridge** [comment]
+1

**Comment by woodbridge** [comment]
Can confirm that this issue still occurs

**Comment by dianaKovalBetterme** [comment]
are there any updates?

**Comment by gnoff** [comment]
@Innei Can you describe what you were doing that this escaping is a problem? Is there some metadata parser that cannot handle HTML escapes?

If you visit your example page you'll see that the browser correctly unescapes the sequence and you can treat the string as if it has an ampersand just fine.

By the way this is not something that Next.js is doing this is a behavior that React does and has done for a very long time. If React did not escape attributes when SSR'ing it would be very easy for anyone rendering user provided strings to be exposed to HTML injection attacks.

Have you (or anyone else in this thread) run into a practical issue where this escaping is breaking some tool?

**Comment by gnoff** [comment]
If you are willing would you mind pointing me to a page hosting an OG image that fails b/c of this escape including which platform fails to fetch the image?

I'm curious because React has SSR'd these content attributes in meta tags exactly this way for years and I can't imagine OG images fail to be fetchable across every SSR'd react project. I wonder if the problem is actually more narrowly triggered either by specific tools or by something else happening in Next.js like double-escaping

**Comment by mpereira** [comment]
Hi @gnoff. I created a minimal reproduction of this issue here: https://github.com/mpereira/react-ssr-meta-tags-escape-bug

I'll echo below the details written in the README.

Since Next.js 13 it is possible to add Open Graph metadata via the [`generateMetadata`](https://nextjs.org/docs/app/api-reference/functions/generate-metadata) function in a `page.tsx` file. In this project, `app/page.tsx` looks like the following:

```typescript
const BASE_URL = "https://webhook.site/a4b648dd-64ad-481b-ac6e-3893ca157b1b";

export async function generateMetadata(_props) {
  const ogImage = `${BASE_URL}?foo=1&bar=2`;

  return {
    openGraph: {
      images: [
        {
          url: ogImage,
        },
      ],
    },
    twitter: {
      images: [ogImage],
    },
  };
}

export default function Home() {
  return (
    <main>
      <h1>react-ssr-meta-tags-escape-bug</h1>
    </main>
  );
}
```

Starting a development server with `npm run dev` and accessing `view-source:http://localhost:3000` will show the following Open Graph tags:

```html
<meta
  property="og:image"
  content="https://webhook.site/a4b648dd-64ad-481b-ac6e-3893ca157b1b?foo=1&amp;bar=2"
/>
<meta
  name="twitter:image"
  content="https://webhook.site/a4b648dd-64ad-481b-ac6e-3893ca157b1b?foo=1&amp;bar=2"
/>
```

The query string separator (`?`) is correctly not escaped in the resulting string, while the query string parameter separator (`&`) is incorrectly escaped to `&amp;`.

This results in a URL that won't behave as expected.

**Comment by mpereira** [comment]
Deployed to Vercel: `view-source:https://react-ssr-meta-tags-escape-bug.vercel.app`

**Comment by Innei** [comment]
> @Innei Can you describe what you were doing that this escaping is a problem? Is there some metadata parser that cannot handle HTML escapes?
> 
> If you visit your example page you'll see that the browser correctly unescapes the sequence and you can treat the string as if it has an ampersand just fine.
> 
> By the way this is not something that Next.js is doing this is a behavior that React does and has done for a very long time. If React did not escape attributes when SSR'ing it would be very easy for anyone rendering user provided strings to be exposed to HTML injection attacks.
> 
> Have you (or anyone else in this thread) run into a practical issue where this escaping is breaking some tool?

As @mpereira said above, we need to keep the & character in the query because when generating the OG we need to determine what type of image to use based on the query's parameters. Obviously, we need to provide a way to not escape characters because here we are ensuring that the information in the query is safe from the risk of XSS.

**Comment by damian-balas** [comment]
Same here :(

`    "next": "14.1.2",`

**Comment by hcorreia** [comment]
@gnoff I created a example site that mimics what our production site did before the workaround.

Its basically what @mpereira already demonstrated.

Website: https://next-metadata-url-issue.vercel.app/
Code: https://github.com/hcorreia/next-metadata-url-issue/blob/main/src/app/page.tsx

Just a note, I think Facebook changed something in their scraper because now it's recognizing the image even with the `&amps;` but I can't prove it.

**Comment by bashidagha** [comment]
this problem still persist

**Comment by charlescrain** [comment]
Yeah experiencing this same issue. Def a problem when using [ImgIx](https://www.imgix.com/) to resize images since it's completely param based

**Comment by jyxjjj** [comment]
It is easy to reproduce: `<meta name="description" content="i am a 'description' content.">`
Then you will view: `i am a &#27;description&#27; content.`

**Comment by jyxjjj** [comment]
Both in https://github.com/facebook/react/issues/13838 especially https://github.com/facebook/react/issues/13838#issuecomment-1241593329

**Comment by ztanner** [comment]
Since this is related to how React renders meta tags, I'm going to close this issue in favor of the one linked above.
